lightning/src/ln/bolt11_payment.rs

```diff
@@ -1,216 +0,0 @@
-		payment_params = payment_params.with_bolt11_features(features.clone()).unwrap();
-	}
-
-	let route_params = RouteParameters::from_payment_params_and_value(payment_params, amount_msat);
-	(payment_hash, recipient_onion, route_params)
-}
-
-#[cfg(test)]
-mod tests {
-	use super::*;
-	use crate::routing::router::Payee;
-	use crate::types::payment::PaymentSecret;
-	use bitcoin::hashes::sha256::Hash as Sha256;
-	use bitcoin::secp256k1::{PublicKey, Secp256k1, SecretKey};
-	use lightning_invoice::{Currency, InvoiceBuilder};
-	use std::time::SystemTime;
-
-	#[test]
-	fn invoice_test() {
-		let payment_hash = Sha256::hash(&[0; 32]);
-		let private_key = SecretKey::from_slice(&[42; 32]).unwrap();
-		let secp_ctx = Secp256k1::new();
-		let public_key = PublicKey::from_secret_key(&secp_ctx, &private_key);
-
-		let timestamp = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap();
-		let invoice = InvoiceBuilder::new(Currency::Bitcoin)
-			.description("test".into())
-			.payment_hash(payment_hash)
-			.payment_secret(PaymentSecret([0; 32]))
-			.duration_since_epoch(timestamp)
-			.min_final_cltv_expiry_delta(144)
-			.amount_milli_satoshis(128)
-			.build_signed(|hash| secp_ctx.sign_ecdsa_recoverable(hash, &private_key))
-			.unwrap();
-
-		assert!(payment_parameters_from_variable_amount_invoice(&invoice, 42).is_err());
-
-		let (hash, onion, params) = payment_parameters_from_invoice(&invoice).unwrap();
-		assert_eq!(&hash.0[..], &payment_hash[..]);
-		assert_eq!(onion.payment_secret, Some(PaymentSecret([0; 32])));
-		assert_eq!(params.final_value_msat, 128);
-		match params.payment_params.payee {
-			Payee::Clear { node_id, .. } => {
-				assert_eq!(node_id, public_key);
-			},
-			_ => panic!(),
-		}
-	}
-
-	#[test]
-	fn zero_value_invoice_test() {
-		let payment_hash = Sha256::hash(&[0; 32]);
-		let private_key = SecretKey::from_slice(&[42; 32]).unwrap();
-		let secp_ctx = Secp256k1::new();
-		let public_key = PublicKey::from_secret_key(&secp_ctx, &private_key);
-
-		let timestamp = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap();
-		let invoice = InvoiceBuilder::new(Currency::Bitcoin)
-			.description("test".into())
-			.payment_hash(payment_hash)
-			.payment_secret(PaymentSecret([0; 32]))
-			.duration_since_epoch(timestamp)
-			.min_final_cltv_expiry_delta(144)
-			.build_signed(|hash| secp_ctx.sign_ecdsa_recoverable(hash, &private_key))
-			.unwrap();
-
-		assert!(payment_parameters_from_invoice(&invoice).is_err());
-
-		let (hash, onion, params) =
-			payment_parameters_from_variable_amount_invoice(&invoice, 42).unwrap();
-		assert_eq!(&hash.0[..], &payment_hash[..]);
-		assert_eq!(onion.payment_secret, Some(PaymentSecret([0; 32])));
-		assert_eq!(params.final_value_msat, 42);
-		match params.payment_params.payee {
-			Payee::Clear { node_id, .. } => {
-				assert_eq!(node_id, public_key);
-			},
-			_ => panic!(),
-		}
-	}
-
-	#[test]
-	fn payment_metadata_end_to_end() {
```

Comment:
Seems like we should keep this test (moving it), but use the new API and also cover the zero-amount cases.

Reply:
Makes sense! I moved the test into our codebase in [pr3617.04](https://github.com/shaavan/rust-lightning/commits/pr3617.04)
Also, introduced a variation on it with `invoice_with_no_amount` case.
Thanks so much for the pointer!

Reply:
Could you also test the failure cases? (i.e., when returning `Bolt11PaymentError::InvalidAmount`)?

Reply:
Done! [pr3617.05](https://github.com/shaavan/rust-lightning/commits/pr3617.05)
Thanks! ✨